api/src/services/purged-docs.js

```diff
@@ -51,59 +65,109 @@ const getPurgedIdsFromChanges = result => {
-    if (cached) {
-      cache.ttl(cacheKey);
-      return Promise.resolve(cached);
+    const cachedIds = await getCachedIds(cacheDocId);
+    if (cachedIds) {
+      return Promise.resolve(cachedIds);
     }
   }
 
   const ids = docIds.map(purgingUtils.getPurgedId);
   let purgeDb;
-  // requesting _changes instead of _all_docs because it's roughly twice faster
-  return getPurgeDb(userCtx.roles)
-    .then(tempDb => purgeDb = tempDb)
-    .then(() => purgeDb.changes({ doc_ids: ids, batch_size: ids.length + 1, seq_interval: ids.length }))
-    .then(result => {
-      purgeIds = getPurgedIdsFromChanges(result);
-      cache?.set(cacheKey, purgeIds);
-      db.close(purgeDb);
-    })
-    .catch(err => catchDbNotFoundError(err, purgeDb))
-    .then(() => purgeIds);
+
+  try {
+    purgeDb = await getPurgeDb(userCtx.roles);
+    // requesting _changes instead of _all_docs because it's roughly twice faster
+    const changesResult = await purgeDb.changes({ doc_ids: ids, batch_size: ids.length + 1, seq_interval: ids.length });
+    purgeIds = getPurgedIdsFromChanges(changesResult);
+    useCache && await setCachedIds(cacheDocId, purgeIds);
+    db.close(purgeDb);
+  } catch (err) {
+    catchDbNotFoundError(err, purgeDb);
+  }
+
+  return purgeIds;
 };
 
 const getUnPurgedIds = (userCtx, docIds) => {
+  docIds.sort();
   return getPurgedIds(userCtx, docIds).then(purgedIds => _.difference(docIds, purgedIds));
 };
 
-const listen = (seq = 'now') => {
-  db.sentinel
-    .changes({ live: true, since: seq })
-    .on('change', change => {
-      seq = change.seq;
-      if (change.id.startsWith('purgelog:') && change.changes[0].rev.startsWith('1-')) {
-        clearCache();
-      }
-    })
-    .on('error', err => {
-      logger.error('Error watching sentinel changes, restarting: %o', err);
-      listen(seq);
-    });
+const listen = () => {
+  watchSentinel();
+  watchUsers();
+  watchMedic();
+  watchSettings();
+};
+
+const watchSentinel = () => {
+  dbWatcher.sentinel(change => {
+    if (change.id.startsWith('purgelog:') && change.changes[0].rev.startsWith('1-')) {
+      return clearCache();
+    }
+  });
+};
+
+const watchUsers = () => {
+  dbWatcher.users(({ id }) => {
+    if (id.startsWith(USER_DOC_PREFIX)) {
+      const name = id.replace(USER_DOC_PREFIX, '');
+      return clearCache(name);
+    }
+  });
+};
+
+const watchMedic = () => {
+  dbWatcher.medic(({ id }) => {
+    if (id.startsWith(USER_DOC_PREFIX)) {
+      const name = id.replace(USER_DOC_PREFIX, '');
+      return clearCache(name);
+    }
+  });
+};
+
+const watchSettings = () => {
+  const key = 'district_admins_access_unallocated_messages';
+  let district_admins_access_unallocated_messages = config.get(key);
+  configWatcher.watch(() => {
+    const newConfigValue = config.get(key);
+    if (newConfigValue !== district_admins_access_unallocated_messages) {
+      district_admins_access_unallocated_messages = newConfigValue;
+      clearCache();
+    }
+  });
```

Comment:
This feels like something the service should provide, like...

```
configWatcher.watch('district_admins_access_unallocated_messages', () => {...});
```

Reply:
This is a cool idea! But it would add some complexity - each time settings are updated, we'd need to iterate over all properties and emit for every change. Do you think it's worth to add this now?

Reply:
Happy to leave it for now.